Date math support for aliases

Tommmster opened the pull request:
Related to issue https://github.com/elastic/elasticsearch/issues/20367

* Does not support date math expressions in the PUT _template API ( see `MetadataIndexTemplateService#validate`)

* When adding an alias

```
curl  -u 'elastic:password'  -XPUT   'http://localhost:9200/regular_index/'
curl  -u 'elastic:password'  -XPOST 'http://localhost:9200/regular_index/_alias/%3Calias-%7Bnow%2FM%7D%3E'
```


* When creating indices
```
# Create <precise-alias-{now/m{dd.HH.mm.ss.SSS}}> for index  <precise-{now/m{dd.HH.mm.ss.SSS}}>
curl -H'content-type:application/json' -u 'elastic:password'  -XPUT 'http://localhost:9200/%3Cprecise-%7Bnow%2Fm%7Bdd.HH.mm.ss.SSS%7D%7D%3E -d \
 '{"aliases" : {"regular_alias":{}, "<precise-alias-{now/m{dd.HH.mm.ss.SSS}}>": {}}}'

curl  -u 'elastic:password'  'http://localhost:9200/_cat/aliases?v'
alias                         index                   filter routing.index routing.search is_write_index
precise-alias-11.02.25.00.000 precise-11.02.25.00.000 -      -             -              -
regular_alias                 precise-11.02.25.00.000 -      -             -              -
not_regular-2021.01.01        index-2021.01.01        -      -             -              -
regular_alias                 index-2021.01.01        -      -             -              -
```



<!--
Thank you for your interest in and contributing to Elasticsearch! There
are a few simple things to check before submitting your pull request
that can help with the review process. You should delete these items
from your submission, but they are here to help bring them to your
attention.
-->

- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)? Yes
- Have you followed the [contributor guidelines](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md)?
- If submitting code, have you built your formula locally prior to submission with `gradle check`?
- If submitting code, is your pull request against master? Unless there is a good reason otherwise, we prefer pull requests against master and will backport as needed. Yes
- If submitting code, have you checked that your submission is for an [OS and architecture that we support](https://www.elastic.co/support/matrix#show_os)?
- If you are submitting this code for a class then read our [policy](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#contributing-as-part-of-a-class) for that.

Tommmster committed:
resolve math expressions for POST _alias and aliases in index settings at creation time

Tommmster committed:
all aliases are set at the same timestamp

Tommmster commented:
Not sure why the CLA check is failing

Tommmster committed:
simple test

Tommmster committed:
small test to trigger CLAs

Tommmster committed:
Revert "small test to trigger CLAs"

This reverts commit 476cf04de233bd58988b027fbc65149964e8e0d9.

Tommmster committed:
test

Tommmster committed:
Adding tests

Tommmster committed:
aliases from index template

elasticmachine commented:
Pinging @elastic/es-core-features (Team:Core/Features)

jtibshirani commented:
@Tommmster sorry for the slow response, this PR slipped through the cracks. I now alerted the relevant team, hopefully they're able to take a look.

martijnvg commented:
@Tommmster Thanks for opening this pr! Apologies for the slow response here. Did you sign the CLA with the same email address used in the commits of this PR?

mark-vieira commented:
@elasticmachine update branch

Tommmster commented:
@martijnvg Yes, the one set in `user.email`. Also tried with the username and (at)Tommmster

tvernum commented:
@Tommmster Those commits don't have a valid email address in them.
You have used `username@gmail` without the trailing `.com`

See https://github.com/elastic/elasticsearch/commit/ee17ceadc1050d6cf95193e31ff115b775c13325.patch for an example.

Tommmster commented:
Thanks @tvernum , seems that was it :)

martijnvg commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:
```diff
@@ -443,6 +443,26 @@ public String resolveDateMathExpression(String dateExpression) {
+     * @return If the specified string is data math expression then this method returns the resolved expression and the instant at which
+     * it was resolved
+     */
+    public ResolvedContext resolveDateMathExpressionAndInstant(String dateExpression) {
```
This method and `ResolvedContext` class seems like not really needed?
The `resolveDateMathExpressionAndInstant(...)` method is only in `TransportCreateAction` and
there the `resolveDateMathExpressionAtInstant(...)` could be used? (using `System.currentTimeMillis()`,
and return value of `resolveDateMathExpressionAtInstant (...)` method)

Tommmster replied:
The rationale behind using the `ResolvedContext` instance is to create the index and aliases set in the index settings with the same timestamps, instead of using `System.currentTimeMillis()` for each

Tommmster replied:
That would work, yes.

martijnvg commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:
```diff
@@ -443,6 +443,26 @@ public String resolveDateMathExpression(String dateExpression) {
+     * @param time instant to consider when parsing the expression
+     * @return If the specified string is data math expression then this method returns the resolved expression.
+     */
+    public String resolveDateMathExpressionAtInstant(String dateExpression, long time) {
```
Maybe just name this method `resolveDateMathExpression`? So that it just overloads `resolveDateMathExpression (...)` method.

martijnvg reviewed (commented):
@Tommmster I took an initial look at your pr. Can you also update the pr to latest master and resolve the merge conflicts?

Tommmster committed:
merged with master, solved conflicts

Tommmster committed:
rename resolveDateMathExpressionAtInstant to resolveDateMathExpression

martijnvg commented on server/src/main/java/org/elasticsearch/action/admin/indices/create/TransportCreateIndexAction.java:
```diff
@@ -36,6 +36,7 @@
 /**
  * Create index action.
  */
+@SuppressWarnings("checkstyle:LineLength")
```
remove this annotation and adjust the line lenght?

martijnvg commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:
```diff
@@ -802,6 +806,46 @@ static int getIndexNumberOfRoutingShards(Settings indexSettings, @Nullable Index
+     * @return the list of resolved aliases, with the explicitly provided aliases occurring first (having a higher priority) followed by
+     * the ones inherited from the templates
+     */
+    public static List<AliasMetadata> resolveAndValidateAliases(String index, Set<Alias> aliases,
```
I think the additional logic of this method at line 827-828 and 837 to 841 can be inlined into the existing `resolveAndValidateAliases(...)` method?

Tommmster replied:
This was a step in the right direction. Removed the overload and inlined the expression resolver

martijnvg commented on server/src/main/java/org/elasticsearch/action/admin/indices/alias/TransportIndicesAliasesAction.java:
```diff
@@ -110,8 +110,12 @@ protected void masterOperation(Task task, final IndicesAliasesRequest request, f
                 switch (action.actionType()) {
                 case ADD:
+
+                    long now = System.currentTimeMillis();
```
I think `long now = System.currentTimeMillis();` should be put before line 110? Otherwise aliases created by an update aliases request can still have different names?

martijnvg reviewed (commented):
Thanks @Tommmster, I left some more comments.

martijnvg commented:
@elasticmachine test this please

Tommmster committed:
Single resolveAndValidateAliases method in MetadataCreateIndexService

Removed IndexNameExpressionResolver.ResolvedContext inner class, using existing overload for IndexNameExpressionResolver#resolveDateMathExpression

Ensure all alias in template have the same creation time (which is the process time)

martijnvg commented:
@elasticmachine test this please

martijnvg commented on docs/reference/indices/add-alias.asciidoc:
```diff
@@ -138,4 +138,5 @@ PUT /logs_20302801
     }
   }
 }
+
```
unneeded?

martijnvg commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:
```diff
@@ -1077,6 +1077,7 @@ private static void validateCompositeTemplate(final ClusterState state,
                                                   final ComposableIndexTemplate template,
                                                   final IndicesService indicesService,
                                                   final NamedXContentRegistry xContentRegistry) throws Exception {
+
```
unneeded?

martijnvg commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:
```diff
@@ -1107,6 +1108,7 @@ private static void validateCompositeTemplate(final ClusterState state,
                 .build())
             .build();
         final IndexMetadata tmpIndexMetadata = stateWithIndex.metadata().index(temporaryIndexName);
+
```
unneeded?

martijnvg commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:
```diff
@@ -1306,6 +1310,7 @@ private void validate(String name, @Nullable Settings settings, List<String> ind
             throw new InvalidIndexTemplateException(name, validationException.getMessage());
         }
 
+
```
unneeded?

martijnvg commented on server/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -525,18 +525,18 @@ protected Node(final Environment initialEnvironment,
-                xContentRegistry,
-                systemIndices,
-                forbidPrivateIndexSettings
+                    settings,
```
unneeded?

martijnvg commented on server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateServiceTests.java:
```diff
@@ -80,6 +82,12 @@
 
 public class MetadataIndexTemplateServiceTests extends ESSingleNodeTestCase {
 
+    private static IndexNameExpressionResolver indexNameExpressionResolver;
```
unneeded changes in this file? This field doesn't seem to be used by the tests?

martijnvg reviewed (commented):
Thanks @Tommmster, I left a number of small comments about changes that are unneeded. Also this build fails with checkstyle failures:

```
[ant:checkstyle] [ERROR] /dev/shm/elastic+elasticsearch+pull-request-1/server/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java:2189: Line is longer than 140 characters (found 141). [LineLength] |  
-- | --
  | [ant:checkstyle] [ERROR] /dev/shm/elastic+elasticsearch+pull-request-1/server/src/test/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexServiceTests.java:602: Line is longer than 140 characters (found 144). [LineLength] |  
  | [ant:checkstyle] [ERROR] /dev/shm/elastic+elasticsearch+pull-request-1/server/src/test/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexServiceTests.java:626: Line is longer than 140 characters (found 156). [LineLength] |  
  | [ant:checkstyle] [ERROR] /dev/shm/elastic+elasticsearch+pull-request-1/server/src/test/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexServiceTests.java:681: Line is longer than 140 characters (found 156). [LineLength] |  
  | [ant:checkstyle] [ERROR] /dev/shm/elastic+elasticsearch+pull-request-1/server/src/test/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexServiceTests.java:710: Line is longer than 140 characters (found 156). [LineLength]
```

If your run `./gradlew -p server precommit` locally then you will see this failures as well.

Tommmster committed:
formatting OK

martijnvg commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into date_math_support_for_aliases

martijnvg commented:
@elasticmachine test this please

martijnvg committed:
Merge remote-tracking branch 'es/master' into date_math_support_for_aliases

martijnvg committed:
undo unrelated changes.

martijnvg committed:
added rest tests

martijnvg commented:
@elasticmachine test this please

martijnvg commented:
Thanks @Tommmster, this looks good. I did update yaml tests to use date math in aliases. I will merge this later today.

Tommmster commented:
> Thanks @Tommmster, this looks good. I did update yaml tests to use date math in aliases. I will merge this later today.

Thank you @martijnvg . How should we proceed with the docs ? I was thinking of doing a separate PR for them.

martijnvg commented:
> How should we proceed with the docs ? We can add them to a separate PR if needed.

Maybe add a small section after `Write index` section in the `docs/reference/indices/aliases.asciidoc` file?
Explaining that date math can be used in aliases?

This can be done in this PR or a follow up, whatever you prefer.

martijnvg committed:
Merge remote-tracking branch 'es/master' into date_math_support_for_aliases

martijnvg committed:
adjust index names to not collide with templates with data streams

martijnvg commented:
@elasticmachine test this please

martijnvg committed:
set yaml skip version to avoid running tests against node that don't support date math in aliases.

martijnvg committed:
fixed test compile error

martijnvg commented:
@elasticmachine test this please

martijnvg commented:
Thanks @Tommmster for working on adding date math support to aliases! I will backport this change to the 7.x branch, so that it will be included in the 7.13 release.

Tommmster commented:
> Thanks @Tommmster for working on adding date math support to aliases! I will backport this change to the 7.x branch, so that it will be included in the 7.13 release.

Thank you @martijnvg , this was a lot of fun. Can't wait for the next one :)
